// @regexper/render@1.0.0/lib/extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExtensions = void 0;
var repeat_any_1 = require("./repeat-any");
var repeat_optional_1 = require("./repeat-optional");
var repeat_required_1 = require("./repeat-required");
var repeat_spec_1 = require("./repeat-spec");
function getExtensions() {
    return {
        Root: { type: 'root' },
        Regexp: { type: 'regexp' },
        Match: { type: 'match' },
        MatchFragment: { type: 'match-fragment' },
        Repeat: { type: 'repeat' },
        RepeatSpec: repeat_spec_1.RepeatSpec,
        RepeatAny: repeat_any_1.RepeatAny,
        RepeatOptional: repeat_optional_1.RepeatOptional,
        RepeatRequired: repeat_required_1.RepeatRequired,
        Anchor: { type: 'anchor' },
        Subexp: { type: 'subexp' },
        Charset: { type: 'charset' },
        CharsetRange: { type: 'charset-range' },
        CharsetEscape: { type: 'charset-escape' },
        Escape: { type: 'escape' },
        Literal: { type: 'literal' },
        AnyCharacter: { type: 'any-character' },
    };
}
exports.getExtensions = getExtensions;
